### DISCUSSION TITLE
Any testers out there? [Customized progress text]

### DISCUSSION THREAD

**Discussion by Dendraspis** [staxrip/staxrip#480] [Ideas]
Hey @44vince44 and @JJKylee ,
hello everybody,

are you interested in testing a new feature? 😉
It's one nobody needs except me - ***again***. 🤣

The current x265 mod version has expanded the progress text with some nice additions like *elapsed time*. The downside is you need to widen the processing window a lot to get/see every information. 

So I decided to make it ... different ..... especially shorter.
The result can be previewed using this exe and the last x265 version by DJATOM `x265 3.4+70`:
~~***StaxRip v2.1.7.5D_Progress9.zip***~~

This *feature* has to be enabled in ***Settings » User Interface*** first and it's only working with x265 at the moment.
A screenshot of the resulting progress text was not made on purpose.

***Do you think this could or should be implemented in StaxRip?***

**Comment by ghost** [comment]
@Dendraspis 

> It's one nobody needs except me - again. 🤣

Nope!!! I was about to ask a new feature for the same matter, but I didn't because I was testing if I would eventually be used to it as it is now !! 😁 
I'll test it of course!!

**Reply by Dendraspis** [reply]
🤣   I didn't need testing, because on a 1080p display there is no space left for the jobs list. So I had to do something - but again I thought it would be just a modification I would only implement locally.  😄

**Comment by ghost** [comment]
@Dendraspis 

That works great !!! 🎉 🚀 

> Do you think this could or should be implemented in StaxRip?

**YES!!!!!**

**Reply by Dendraspis** [reply]
🤣 

Do you have any suggestions or questions? Can I improve anything? Is it understandable? 
I tried to be as short as possible without losing readability.  🤔 

I could save some more space by removing spaces between numbers and units, e.g. `1.08fps` or `1234.56kb/s` but I think that would affect the readability too much.  🤔

**Reply by ghost** [reply]
@Dendraspis It is very readable, but I have a small suggestion for consistence:
The remaining time should appear with a minus before it, otherwise the user could think it is the total time (some programs like Virtualdub used to give elapsed time and total time). So it should appear like that:
84MB (245MB), 00:01:56  (-01:20:54)  
This adds only one character.
For the other data, please change nothing 😄 indeed, as you say it could make information not understandable!

**Reply by Dendraspis** [reply]
Okay, minus is applied.  👍 

What about removing the decimal places in estimated size?  `123.45 MB` » `123 MB`
Honestly I don't think that anybody would need that information, especially when it changes every second.  🤔 

Here you can find both ideas applied:
[StaxRip v2.1.7.5D_Progress2.zip](https://github.com/staxrip/staxrip/files/5868489/StaxRip.v2.1.7.5D_Progress2.zip)

**Reply by ghost** [reply]
Oh yes, I agree with you completely ! It is an estimate ;-) 

I have downloaded it and tested it: it looks great!!! Thank you @Dendraspis 👍 👍 

Now I can finally reduce the size of the window 😁

**Comment by JJKylee** [comment]
![StaxRip_2 1 7 5_new_x265_progress_text_1_20210125](https://user-images.githubusercontent.com/3157518/105772900-c3468f80-5f17-11eb-838a-e356a332a220.jpg)

![StaxRip_2 1 7 5_new_x265_progress_text_2_20210125](https://user-images.githubusercontent.com/3157518/105772920-c8a3da00-5f17-11eb-835e-105a006f67fc.jpg)

It's **as good as it can be**. 👍
Now I can see that deep down in my heart I've always been yearning for this kind of feature!!! 😆
Please apply this to other encoders too. 🙏😅

**Reply by Dendraspis** [reply]
Just a side note:
The part at the beginning `[ 2.6%]` was also re-designed.  😁 

Seems like this feature is very important for v2.1.7.6.  😂 

> Please apply this to other encoders too. 🙏😅

The problem is that only this x265 mod gives us all nice information. But I also implemented it for x264 - without *elapsed* obviously.

**Reply by Patman86** [reply]
> The problem is that only this x265 mod gives us all nice information. But I also implemented it for x264 - without *elapsed* obviously.

I think the x264 version from DJATOM will have a progress bar similar to that of  x265.

[enc time x264 aMod](https://github.com/DJATOM/x264-aMod-patches/blob/master/03-enc-time.diff)

**Reply by Dendraspis** [reply]
I hope so. Meanwhile we can use the current stripped down version. 😁

@JJKylee 
Seriously, it looks different.  🤓

**Comment by JJKylee** [comment]
Unfortunately, `StaxRip.v2.1.7.5D_Progress2.zip` seems to have a critical issue with other encoders.
(`NVEnc` with `qaac`)

```
------------------------------ Exception ------------------------------

System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'LabelProgressBar'.
   at System.Windows.Forms.Control.CreateHandle()
   at System.Windows.Forms.Control.get_Handle()
   at System.Windows.Forms.Control.get_WindowText()
   at System.Windows.Forms.Control.set_WindowText(String value)
   at System.Windows.Forms.Control.set_Text(String value)
   at StaxRip.UI.LabelProgressBar.set_Text(String Value)
   at StaxRip.ProcController._Closure$__44-0._Lambda$__0()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at StaxRip.ProcController.VB$StateMachine_44_SetText.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```

The full log is here:
[test_1audio_staxrip.log](https://github.com/staxrip/staxrip/files/5877285/test_1audio_staxrip.log)

[StaxRip v2.1.7.5D_NVEnc.zip](https://github.com/staxrip/staxrip/files/5876786/StaxRip.v2.1.7.5D_NVEnc.zip) also has the same problem.
See [here](https://github.com/staxrip/staxrip/issues/468#issuecomment-767950278).

I'm sad that I have to roll back. 😭 The original 2.1.7.5 executable does not have this issue.

**Reply by ghost** [reply]
I had that with x265 as well, only once. It happened somehow when doing a drag and drop on the window of a new video after creating previous job. But it happened only once and I couldn't reproduce ever (BTW, I only use x265)

**Reply by Dendraspis** [reply]
Okay, fixing will be no problem. 
Thanks for reporting, I couldn't force any exceptions while testing and even my normal encodes finished without issues.

Is it possible the problem only occurs at the end of an encode? Seems like using an encoder without this feature is too fast.  😂 

I'll post the fix later...

....and here it is:
[StaxRip v2.1.7.5D_Progress3.zip](https://github.com/staxrip/staxrip/files/5878584/StaxRip.v2.1.7.5D_Progress3.zip)

Now an appearing exception is nearly impossible. 😬  But a little testing would be nice, because I couldn't reproduce it.

**Reply by ghost** [reply]
Got it, i'll test, but provided that I only had that issue only once previously, I'm not sure my tests will be significant. @JJKylee please check !

**Reply by JJKylee** [reply]
> ....and here it is:
> [StaxRip v2.1.7.5D_Progress3.zip](https://github.com/staxrip/staxrip/files/5878584/StaxRip.v2.1.7.5D_Progress3.zip)

The exception error is gone for NVEnc. Good job! 👍

But now there's another problem. x265 encoding progress bar stops intermittently for a few seconds and then rapidly catches up. For a 2 minute test video, this happens a couple times - repeatedly.

I tested the same video with the same x265 settings (`--crf 27`) on the original 2.1.7.5 binary, and I see no stuttering in the progress bar for this one. 😭

The same phenomenon happens with x264 as well, and [StaxRip v2.1.7.5D_NVEnc2.zip](https://github.com/staxrip/staxrip/files/5879580/StaxRip.v2.1.7.5D_NVEnc2.zip) also shows the same stuttering issue.

I'm not sure if it's my own or not. @44vince44, can you test the new binaries with x265 and x264, please?

**Comment by JJKylee** [comment]
I hope this is not too much of a digression, but since we're talking about the progress bar text, maybe this is a proper thread.

While testing @DJATOM's [new x264 build](https://github.com/DJATOM/x264-aMod/releases/tag/r3039%2B16), I've found that the progress bar text cannot pick up well enough.

See this screenshot:

![StaxRip_2 1 7 5_DJATOM_x264_progress_text_20210127](https://user-images.githubusercontent.com/3157518/105996473-3870ac80-605f-11eb-9a7d-2b06707030ab.jpg)

OTOH, the existing x264 build by @Patman86 shows the correct progress bar text:

![StaxRip_2 1 7 5_Patman_x264_progress_text_20210127](https://user-images.githubusercontent.com/3157518/105996584-66ee8780-605f-11eb-92a1-722082607c3d.jpg)

I guess this is due to the different progress report texts of those two builds.
Well, just FYI. 😄

**Reply by Dendraspis** [reply]
Ok, please test both with this one:  😞 
[StaxRip v2.1.7.5D_Progress4.zip](https://github.com/staxrip/staxrip/files/5880516/StaxRip.v2.1.7.5D_Progress4.zip)

**Reply by DJATOM** [reply]
It prints progress header before actual progress lines, so you see that. 
I can add an option to disable header, if parsing and ignoring that line is difficult,

**Reply by Dendraspis** [reply]
@DJATOM 
That would be great in order to customize the output. But also without this option it could look annoying if there is a bigger gap, for example if the window is a lot taller.
I think the header is a nice option using the encoder directly via console - only.

**Reply by Dendraspis** [reply]
@JJKylee 
The output from the x264 build by @DJATOM  is correct - there is no universal solution to rewrite the values The new code checks the running encoder and verifies the given output. If it matches, the output will be customized, otherwise it's not modified.
So when this build will be taken for StaxRip, I have to adjust the match.

**Reply by Dendraspis** [reply]
@DJATOM 
Parsing the output wouldn't be difficult and after a short brainstorming with myself I can make it work, even without an additional parameter in your build.
Let me try it first before you spend time implementing it... sorry for my late response.

**Reply by Dendraspis** [reply]
@DJATOM , @JJKylee 
It works, expect the header, that I can't turn off on condition - would need *some* code rewriting:

![SRx264](https://user-images.githubusercontent.com/50659978/106010230-7f8b8b80-60b9-11eb-969f-86ddc547b141.JPG)

But I would say this is okay for such a feature and x264.

Here the new version:
[StaxRip v2.1.7.5D_Progress5.zip](https://github.com/staxrip/staxrip/files/5881098/StaxRip.v2.1.7.5D_Progress5.zip)

**Reply by JJKylee** [reply]
OK. I'll give it a try when I get back to my computer in several hours. 😄

**Reply by DJATOM** [reply]
I made new x264 release with `--no-progress-header`.

**Reply by Dendraspis** [reply]
At least a very small change for you.  😊
Don't know yet if I can use it. 🤔

**Reply by DJATOM** [reply]
There is a way to determine binary's version:
```
"x264-aMod-x64-core161-r3039+17.exe" --version
x264 0.161.3039+17 c76b8b8
(libswscale 5.8.100)
(libavformat 58.65.101)
(ffmpegsource 2.40.0.0)
(lsmash 2.16.1)
built on Jan 27 2021, gcc: 10.2.1 20200926
x264 configuration: --chroma-format=all
libx264 configuration: --chroma-format=all
x264 license: GPL version 2 or later
libswscale/libavformat/ffmpegsource license: GPL version 3 or later
```
Probably staxrip can rely on that version to actually add `--no-progress-header` option into command (transparently for user).

**Reply by DJATOM** [reply]
Another way to find it also run a help command and search for that option. It's described there.

**Reply by Dendraspis** [reply]
First we have to check if @stax76 is going to include your build as the new default including the new dll parameter. If so and StaxRip will use the new parameter, your build will be the only one that will work with StaxRip, so no need to prove the version. 😁
It's like with x265. 😎

**Reply by JJKylee** [reply]
I tested @DJATOM's [x264-aMod-x64-core161-r3039+17.7z](https://github.com/DJATOM/x264-aMod/releases/tag/r3039%2B17) and [x265-x64-v3.4+65-aMod-gcc10.2.1+opt.7z](https://github.com/DJATOM/x265-aMod/releases/tag/3.4%2B65) on [StaxRip v2.1.7.5D_Progress5.zip](https://github.com/staxrip/staxrip/files/5881098/StaxRip.v2.1.7.5D_Progress5.zip) along with @rigaya's [NVEnc 5.25](https://github.com/rigaya/NVEnc/releases/tag/5.25).

## [x264]

Without `--no-progress-header`:

![StaxRip v2 1 7 5D_Progress5_DJATOM_x264-aMod-x64-core161-r3039+17_progress_text_20210127](https://user-images.githubusercontent.com/3157518/106067642-f6be2100-60b3-11eb-9530-75b8f266d19b.jpg)

With `--no-progress-header` as a custom option:

![StaxRip v2 1 7 5D_Progress5_DJATOM_x264-aMod-x64-core161-r3039+17_progress_text_no-header_20210127](https://user-images.githubusercontent.com/3157518/106067674-03db1000-60b4-11eb-91b4-2e545eda3028.jpg)

Well, I like this simplicity. 😄

## [x265]

![StaxRip v2 1 7 5D_Progress5_DJATOM_x265-x64-v3 4+65-aMod-gcc10 2 1_progress_text_20210127](https://user-images.githubusercontent.com/3157518/106067720-1d7c5780-60b4-11eb-9f33-212bfed34997.jpg)

No stuttering in the progress bar any more! Good job, @Dendraspis. 😃

BTW, is it normal that `--reader-options library=...` option appears even for installed AviSynth?

## [NVEnc]

No more exception error!
And the new layout for options is very helpful. 👍

**Reply by Dendraspis** [reply]
> No stuttering in the progressive bar any more! Good job, @Dendraspis. 😃

No *good job*, because I created this bug by myself.  😋

> BTW, is it normal that `--reader-options library=...` option appears even for installed AviSynth?

Is the path correct? 🤔

> No more exception error!
> And the new layout for options is very helpful. 👍

👍🏻👍🏻

And you didn't enable the customized progress text to see, if it's working. 😝😂

**Reply by JJKylee** [reply]
> > BTW, is it normal that `--reader-options library=...` option appears even for installed AviSynth?
> 
> Is the path correct? 🤔

Yes, it's correct.

> And you didn't enable the customized progress text to see, if it's working. 😝😂

Right. I'm afraid of doing so, but well, let me give a try anyway. 😝

**Reply by JJKylee** [reply]
I turned on the customized progress text option, and everything seems OK - no exception errors, no stuttering issues, etc.
Furthermore, x264 with `--no-progress-header` option yields this neat progress report:

![StaxRip v2 1 7 5D_Progress5_DJATOM_x264-aMod-x64-core161-r3039+17_progress_text_no-header_new-progress-bar_20210127](https://user-images.githubusercontent.com/3157518/106070999-57505c80-60ba-11eb-8b35-5a872e71c186.jpg)

I think I'm gonna like it a lot. 😄

**Reply by Dendraspis** [reply]
Yeah, the issues are gone, really.  🤓
Thanks for testing.  

I'm curious when someone asks to customize the NVEMC progress.  😂

**Comment by JJKylee** [comment]
@DJATOM released a new `x265` build today ([3.4+70-aMod](https://github.com/DJATOM/x265-aMod/releases/tag/3.4%2B70)) and he introduced two new options: `--progress-readframes` and `--reader-options use-script-sar=1`. The latter is for mimicking `vspipe`, so it's not related to this thread, but I was concerned that the first would cause an issue with @Dendraspis's customized progress text. And my testing proved me right.

- StaxRip: [StaxRip v2.1.7.5D_Warp.zip](https://github.com/staxrip/staxrip/files/5903783/StaxRip.v2.1.7.5D_Warp.zip) (probably the most recent build before 2.1.7.6 as of now)
- x265: 3.4+70-aMod

**[No `--progress-readframes` custom option]**

It's the default and the customized progress text works well.

![StaxRip_2 1 7 6-pre_x265-3 4+70-aMod_customized-progress-text_20210202](https://user-images.githubusercontent.com/3157518/106638110-5abf6a00-6538-11eb-9f6f-b2bc9e9e2f8a.jpg)

**[`--progress-readframes`]**

The customized progress text breaks down.

![StaxRip_2 1 7 6-pre_x265-3 4+70-aMod--progress-readframes_customized-progress-text_20210202](https://user-images.githubusercontent.com/3157518/106638238-82163700-6538-11eb-8618-21484214fbf1.jpg)

It's not a big deal, but... you know... 😓

Sorry, @Dendraspis, for passing a task to you. 😝

**Reply by DJATOM** [reply]
That's why it's not enabled by default 😆 
I thought it will break StaxRip regexp and that's bad.
Also it's probably useless for 99% people and only needed when you want to monitor how much frames passed from vapoursynth/avisynth. Yet it was interesting to implement.

**Reply by Dendraspis** [reply]
@DJATOM
You can set it as default - no problems with that. The only problem is: I have to know what's going on.  😁

It will be supported ... maybe tomorrow. ⌛

**Reply by Dendraspis** [reply]
@JJKylee 
Could you give it a try, please?  Can't test at the moment, so it's a ***blind date***.  😁 

***File removed***

**Reply by JJKylee** [reply]
The customized progress text works good but I see an unnecessary pair of parentheses for both default and `--progress-readframes` cases. See these screenshots:

![StaxRip v2 1 7 5D_Progress7_x265-3 4+70-aMod_default_progress-text_20210202](https://user-images.githubusercontent.com/3157518/106675908-cde1d400-656a-11eb-8fed-4b6912eadf0c.jpg)

![StaxRip v2 1 7 5D_Progress7_x265-3 4+70-aMod_--progress-readframes_progress-text_20210202](https://user-images.githubusercontent.com/3157518/106675981-ebaf3900-656a-11eb-87af-584a8d28a43f.jpg)

I believe you will fix this. 😉

**Reply by Dendraspis** [reply]
🤣    You were too fast and so you have tested ***Progress7***, which was replaced by ***Progress8*** after only a few minutes.  😁 

But that was as expected. ***Progress8*** showed empty parentheses when there was no readframes value.
I'll implement the fix., but if you wanna test (or use) it:

[StaxRip v2.1.7.5D_Progress9.zip](https://github.com/staxrip/staxrip/files/5915042/StaxRip.v2.1.7.5D_Progress9.zip)

Once more:  Thanks for testing @JJKylee  !

**Reply by JJKylee** [reply]
> [StaxRip v2.1.7.5D_Progress9.zip](https://github.com/staxrip/staxrip/files/5915042/StaxRip.v2.1.7.5D_Progress9.zip)

Thanks, but unfortunately this one lacks a lot of info in the progress text.
I'll be happily waiting for your final fix version. 😉

**Reply by Dendraspis** [reply]
😲    I shouldn't sit in front of my computer this evening.  😞  Glad you have responded, because this version was on it's way to the next beta.  🙄 

Should be fixed now....
[StaxRip v2.1.7.5D_Progress10.zip](https://github.com/staxrip/staxrip/files/5915217/StaxRip.v2.1.7.5D_Progress10.zip)

**Reply by JJKylee** [reply]
You're right. It's perfect! 👍 

![StaxRip v2 1 7 5D_Progress10_x265-3 4+70-aMod_default_progress-text_20210202](https://user-images.githubusercontent.com/3157518/106695259-157a5700-658f-11eb-8659-c92be3096a36.jpg)

![StaxRip v2 1 7 5D_Progress10_x265-3 4+70-aMod_--progress-readframes_progress-text_20210202](https://user-images.githubusercontent.com/3157518/106695276-1c08ce80-658f-11eb-8aba-725d59f5f57a.jpg)

Good job, man~ 😄

**Comment by Dendraspis** [comment]
Since v2.1.7.6 `--progress-readframes` will be supported for x265.
Additionally I decided to *kill* (no rounding) the decimal places for the bitrate:  `8261.69 kb/s` » `8261 kb/s`. This saves 3 characters and I can't believe someone cares about less than 1kb in bitrate differences.  👶

**Comment by stax76** [comment]
I've enabled it now by default with a small change, if the regex don't match, it gets disabled.

**Reply by Dendraspis** [reply]
Great, so I can remove this task from my To-Do list.  😄  👍  

By default I wanted this option to be turned off because it needs some calculations, meaning CPU time, and I was not sure, if anybody would like/need it. But it seems like you like this option, too.  😄 
On the other side the progress bar takes a lot more power.  😁

**Comment by stax76** [comment]
I tried to measure it for a while, staxrip vs Windows Terminal, could not find a real difference.

**Reply by Dendraspis** [reply]
You have to think in nanoseconds. 😜 😁

Wasn't meant too serious, and even 100ms would never be noticed.

**Reply by ghost** [reply]
You mean if I have an encode that takes 3 years, it could take a couple of hours more ? this in unacceptable 🤣

**Reply by Dendraspis** [reply]
Maybe not even an hour.  😏 

And maybe I was thinking a bit too extreme, I admit.  😋

**Comment by JJKylee** [comment]
Sorry to toss another job, @Dendraspis, but @Patman86 is building his own binaries again, and he seems to have decided to add his own custom progress texts. 😆

His `x264`(in `x264-0.161.3048-b86ae3c-[Mod-by-Patman]-x64-gcc10.2.0.7z`) looks like this:

![StaxRip_2 1 8 0_Patman_x264_default_progress-text_20210212](https://user-images.githubusercontent.com/3157518/107838714-0009e780-6d5c-11eb-9865-efeb87e1c933.jpg)

FYI, running `x264 --version`...:

```
x264 0.161.3048 b86ae3c [Mod by Patman]
[...]
```

He's working on his own `x265` now, and it'll be out tomorrow. 😄

**Reply by ghost** [reply]
LOOOOOLLLLL 🤣

**Reply by Dendraspis** [reply]
A bit off-topic, but: Which app do you both use to create those *illustrations*? 🤔

@44vince44 
Still laughing? 😁

**Reply by ghost** [reply]
@Patman86 I know we can't ask you to mimic the verbosity output of DJATOM's mods, but it seems very inhumane to ask @Dendraspis to re-format it for each mod 😕 
What's the solution here ???
Your formatting is very clean and nice but it is too large!
And anyway, it's also up to @Dendraspis to decide also and to comment!

**Reply by ghost** [reply]
@Dendraspis mine is called Techsmith Snag-it (it has a screen capture module and an editing module that allows to write and draw boxes, circles, text etc...)

**Reply by Patman86** [reply]
@44vince44 
I make some little changes this evening and follow @Dendraspis changes 😉

**Reply by JJKylee** [reply]
I use SnagIt too. A little heavy, but convenient.

**Reply by ghost** [reply]
@JJKylee I'm using both an old version and the latest version of Snagit. I'm more used to the old capture tool (of V10, which is dated 2010)...

**Reply by Dendraspis** [reply]
Generally I have no problem adding another checking pattern, as long as it's worth doing.
Personally I don't like changing the maintainers with each beta version. @DJATOM introduced direct input, some other nice features and he worked with us, so I think we could give him some credit by using his mod (for a while). 🤔 

@Patman86 
Feel free using a format of your choice - I can add an `If` for you as well. 😊 

@44vince44 
Thanks, my white knight. 😂 

@JJKylee 
Even version checking for x264 is not implemented yet. 🤔  Might take a while...

@44vince44 , @JJKylee 
I know that you use the same app - the text and arrows looked always the same.  😄 
SnagIt - used it a long time ago, too.  😁

**Reply by Dendraspis** [reply]
@Patman86 
I'm still waiting to adjust my patterns. 😄 
Hope your x265 progress looks the same as the one from x264, so I can simply copy&paste it.  😁

**Reply by DJATOM** [reply]
> I think we could give him some credit by using his mod

My forks have gained some stars in those few weeks, thanks. Feel free to update onto Pat's binaries since they will probably use newer codebase (I'll update x265 codebase with 3.5 final and x264 with core 162).

**Reply by Patman86** [reply]
> @Patman86 
> I'm still waiting to adjust my patterns. 😄 
> Hope your x265 progress looks the same as the one from x264, so I can simply copy&paste it.  😁 

I'm still at it, unfortunately ... I should be ready in the next 1-2 days and it will be a mix of Asuna/aMod.  My progress bar will look like this:

` ...\ ... Frames @ ... FPS | ... kB/s | elapsed (-remain) | size (est. size) `

@DJATOM thx for your codes (vpy-support, external library support)

@msg7086 thx for your codes (mp4 and mkv output support)

**Reply by msg7086** [reply]
I would rather using something more machine friendly, like a JSON output, for GUIs, like [this](https://github.com/msg7086/x265-Yuuki-Asuna/commit/80dde77d4dbec3c29a3f611f7d6ea8551254939a).

**Reply by DJATOM** [reply]
That might be a good idea (to avoid parsing completely and use built-in json library on client application to read progress status). I'll consider it with next release (since I'm writing a python wrapper, which parses and exposes `stderr` to console and built-in webserver, allowing remote monitoring feature 😄).

**Reply by msg7086** [reply]
Yea I was writing a remote bangumi auto encoder and I needed that progress update feature for monitoring, I thought maybe having a dedicated progress file would be much more easier for both side.

**Comment by stax76** [comment]
StaxRip is .NET Framework based which has no built-in json parser, .NET Framework was discontinued in favor of .NET Core (now called .NET 5). I usually work around it by using the powershell API, but that is slow. More or less all .NET devs use a component by newtonsoft to parse json, but this library has a large file size, and I don't like external dependencies much, still often use xml instead.

.NET Core would have a json parser built in, but we can't easily migrate to that because staxrip has complex form designer code and the form designer needed a major rewrite done by MS (running out of process now), it's very likely again unstable like often in the past, and .NET Core has other disadvantages, at least at the moment.

Parsing the output is simple:

```
        If value.Contains("%") Then
            value = value.Left("%")

            If value.Contains("[") Then
                value = value.Right("[")
            End If

            If value.Contains(" ") Then
                value = value.RightLast(" ")
            End If

            If value.IsDouble Then
                Dim val = value.ToDouble

                If LastProgress <> val Then
                    ProcForm.Taskbar?.SetState(TaskbarStates.Normal)
                    ProcForm.Taskbar?.SetValue(Math.Max(val, 1), 100)
                    ProcForm.NotifyIcon.Text = val & "%"
                    ProgressBar.Value = val
                    LastProgress = val
                End If

                Exit Sub
            End If
        ElseIf
```

It creates some new string objects, maybe regex would be faster, but it should be neglectable.

The regex wrapper from Dendraspis can be seen here:

https://github.com/staxrip/staxrip/blob/master/General/ProcController.vb#L105

CPU time is fine according to my tests, I couldn't measure a real difference. I think Patman wants to change it directly in x265. I hope a collaboration effort is possible, suggesting sMod (Synth Mod) as name, avisynth and vapoursynth support are the most important Mod feature.

**Reply by msg7086** [reply]
Nothing prevents you from consuming an XML variant of progress file. I work in Ruby so I use JSON, but you can always change the code and compose XML strings. You can split the parameter into progress-json and progress-xml, or you can check the extension of the file and smartly choosing between two.

Regarding the mod name, you can use StaxMod or sMod if it's intended to work best with StaxRip.

**Comment by stax76** [comment]
I think I've not yet understood how the discussion feature works..., there were comments about json, somewhere. 🤔

**Reply by DJATOM** [reply]
You have to load more answers by pressing that link 
![Screenshot_2021-02-15-23-24-16-750_org mozilla firefox](https://user-images.githubusercontent.com/4888694/107994346-22626600-6fe5-11eb-8700-dfffc45ca473.jpg)

**Comment by stax76** [comment]
@Dendraspis

Were all three regex pattern designed for DJATOM builds?

**Reply by Dendraspis** [reply]
Nope, I think it was one of the x265 patterns that was made for the old Patman builds.
Do you want to delete it even now?

**Reply by stax76** [reply]
Since builds can change something might need to be adjusted or optimized.

For that code comments naming build/mod type and example output would be very helpful.

**Reply by Dendraspis** [reply]
I wanted to wait till Patman's new x265 came out - and now if he implements my request.  😁 
But feel free - was just a question.  😁

**Comment by Dendraspis** [comment]
The next build also supports latest x265 3.5 RC1 Patman build.   😁
